external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/HdfsBolt.java

```diff
@@ -88,35 +99,94 @@ public void doPrepare(Map conf, TopologyContext topologyContext, OutputCollector
-            byte[] bytes = this.format.format(tuple);
-            synchronized (this.writeLock) {
-                out.write(bytes);
-                this.offset += bytes.length;
-
-                if (this.syncPolicy.mark(tuple, this.offset)) {
-                    if (this.out instanceof HdfsDataOutputStream) {
-                        ((HdfsDataOutputStream) this.out).hsync(EnumSet.of(SyncFlag.UPDATE_LENGTH));
-                    } else {
-                        this.out.hsync();
-                    }
-                    this.syncPolicy.reset();
+        boolean forceRotate = false;
+        synchronized (this.writeLock) {
+            boolean forceSync = false;
+            if (TupleUtils.isTick(tuple)) {
+                LOG.debug("TICK! forcing a file system flush");
+                forceSync = true;
+            }
+            else {
+                try {
+                    writeAndAddTuple(tuple);
+                } catch (IOException e) {
+                    //If the write failed, try to sync anything already written
+                    LOG.info("Tuple failed to write, forcing a flush of existing data.");
+                    this.collector.reportError(e);
+                    forceSync = true;
+                    this.collector.fail(tuple);
                 }
             }
 
-            this.collector.ack(tuple);
+            if (this.syncPolicy.mark(tuple, this.offset) || forceSync) {
+                try {
+                    syncAndAckTuples();
+                } catch (IOException e) {
+                    LOG.warn("Data could not be synced to filesystem, failing this batch of tuples");
+                    this.collector.reportError(e);
+                    //Force rotation to get a new file handle
+                    forceRotate = true;
+                    for (Tuple t : tupleBatch)
+                        this.collector.fail(t);
+                    tupleBatch.clear();
+                }
+            }
+        }
 
-            if(this.rotationPolicy.mark(tuple, this.offset)){
-                rotateOutputFile(); // synchronized
-                this.offset = 0;
-                this.rotationPolicy.reset();
+        if(this.rotationPolicy.mark(tuple, this.offset) || forceRotate) {
+            try {
+                rotateAndReset();
+            } catch (IOException e) {
+                this.collector.reportError(e);
+                LOG.warn("File could not be rotated");
+                //At this point there is nothing to do.  In all likelihood any filesystem operations will fail.
+                //The next tuple will almost certainly fail to write and/or sync, which force a rotation.  That
+                //will give rotateAndReset() a chance to work which includes creating a fresh file handle.
             }
-        } catch (IOException e) {
-            this.collector.reportError(e);
-            this.collector.fail(tuple);
         }
     }
 
+    private void rotateAndReset() throws IOException {
```

Comment:
You could have a low default (say 10 continuous failures or so) and also add an option for the users to change this. @harshach can you recommend a default number of retries after which the task can exit so that storm would spawn the task in another worker where it might succeed?